core/src/main/java/org/elasticsearch/cluster/routing/allocation/NodeAllocationResult.java

```diff
@@ -19,59 +19,318 @@
+         * returns -1 if unknown.
+         */
+        public long getVersion() {
+            return version;
+        }
+
+        /**
+         * Gets the number of matching bytes the shard copy has with the primary shard.
+         * Returns -1 if not applicable (this value only applies to assigning replica shards).
+         */
+        public long getMatchingBytes() {
+            return matchingBytes;
+        }
+
+        /**
+         * Gets the store exception, if one exists.  Otherwise, {@code null} is returned.  A store
+         * exception will only exist if {@link ShardStore#getStoreStatus()} returns
+         * {@link StoreStatus#CORRUPT} or {@link StoreStatus#IO_ERROR}.
+         */
+        public Exception getStoreException() {
+            return storeException;
         }
-        if (other == null || getClass() != other.getClass()) {
-            return false;
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            storeStatus.writeTo(out);
+            out.writeOptionalString(allocationId);
+            out.writeLong(version);
+            out.writeLong(matchingBytes);
+            if (storeException != null) {
+                out.writeBoolean(true);
+                out.writeException(storeException);
+            } else {
+                out.writeBoolean(false);
+            }
+        }
+
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            builder.startObject("store");
+            {
+                builder.field("status", storeStatus.toString());
+                if (allocationId != null) {
+                    builder.field("allocation_id", allocationId);
+                }
+                if (version > 0) {
+                    builder.field("version", Version.fromId((int) version));
+                }
+                if (matchingBytes >= 0) {
+                    builder.field("matching_bytes", new ByteSizeValue(matchingBytes).toString());
+                }
+                if (storeException != null) {
+                    builder.field("store_exception", ExceptionsHelper.detailedMessage(storeException));
+                }
+            }
+            builder.endObject();
+            return builder;
         }
-        NodeAllocationResult that = (NodeAllocationResult) other;
-        return decision.equals(that.decision) && Float.compare(weight, that.weight) == 0;
     }
 
-    @Override
-    public int hashCode() {
-        return Objects.hash(decision, weight);
+    /** An enum representing the state of the shard store's copy of the data on a node */
+    public enum StoreStatus {
+        // A copy of the data is available on this node
+        AVAILABLE((byte) 0),
+        // The copy of the data on the node is corrupt
+        CORRUPT((byte) 1),
+        // There was an error reading this node's copy of the data
+        IO_ERROR((byte) 2),
```

Comment:
I added the separation between status and readability here: 6a19c185197138410fb21df818a6ef246048d4d5